### ISSUE TITLE
Add support for NATS Jetstream

### ISSUE THREAD

**Issue by mfreeman451** [timeplus-io/proton#535]
### Use Case
Proton/Timeplus is recognized for its high-performance streaming SQL capabilities, largely powered by ClickHouse. To enhance its integration capabilities and messaging support, it would be beneficial to add support for NATS Jetstream. NATS Jetstream offers advanced features like persistence, at-least-once delivery, consumer offsets, and replayability, which are essential for reliable and scalable message streaming in complex, distributed systems. By integrating NATS Jetstream, Proton/Timeplus can be leveraged in more diverse environments, particularly where NATS is already an integral part of the messaging infrastructure.

### Desired Solution
The ideal solution would be the implementation of a NATS Jetstream connector within Proton/Timeplus. This connector should be capable of both publishing to and subscribing from NATS Jetstream topics. Key features would include:

- **Stream Ingestion**: Ability to ingest data streams from NATS Jetstream, enabling real-time analytics on the streaming data.
- **Fault Tolerance**: Ensuring at-least-once delivery semantics to guarantee message processing in the face of network or system failures.
- **Scalability**: Support for NATS Jetstream scalability features, including partitioning and load balancing across multiple consumers.
- **Ease of Configuration**: Simple configuration setup for connecting Proton/Timeplus with a NATS Jetstream server, including stream and consumer configuration.
- **Data Transformation and Enrichment**: Capability to perform SQL-based transformations and enrichments on the ingested data from NATS Jetstream.

### Alternatives Considered
Currently, integrating NATS Jetstream with Proton/Timeplus requires custom solutions or middleware, which can be complex and less efficient. Other messaging systems like Apache Kafka are supported, but they do not align with the infrastructure where NATS Jetstream is already in use.

### Additional Context
The addition of NATS Jetstream support would significantly enhance Proton/Timeplus's appeal in environments where lightweight and high-performance messaging is crucial. It would also align Proton/Timeplus with modern cloud-native architectures that are increasingly adopting NATS Jetstream for messaging.

**Comment by jovezhong** [comment]
Thank you @mfreeman451  to add this feature request. It's certainly something we badly want to add to Proton. Just checking whether you may want to contribute code on this.

**Comment by mfreeman451** [comment]
would love to @jovezhong but C++ is outside of my wheel house.

**Comment by jovezhong** [comment]
No problem at all. Maybe we can invite external contributors to work on this and you can help us to review/test the integration. Let me setup a bounty on this. Our team right now are adding native ClickHouse support and will add more enhancements to the Kafka External Stream. If any one who are familiar with C++ and NATS to pick up this task, that'll be great

**Comment by oliverqx** [comment]
/attempt #535

| [Algora profile](https://console.algora.io/@/oliverqx0) | Completed bounties | Tech | Active attempts | Options |
|---|---|---|---|---|
| @oliverqx | | TypeScript, Python | | [Cancel attempt](https://console.algora.io/api/bounties/cls13uc4c0002ih0g68ihcgc6/cancel-attempt) |

**Comment by hwinkel** [comment]
Has not clickhouse already a NATS integration ?

**Comment by yokofly** [comment]
> Has not clickhouse already a NATS integration ?

no, clickhouse still does not fully support NATS Jetstream I guess?
I find this PR still working in progress: https://github.com/ClickHouse/ClickHouse/pull/51281
full discussion here: https://github.com/ClickHouse/ClickHouse/issues/39459

**Comment by jovezhong** [comment]
We certainly can leverage some code of NATS integration in ClickHouse. Most likely it'll be a special Table Engine in ClickHouse. However in Proton, we prefer reading/writing NATS with External Stream, just like how it [reads/writes Kafka](https://docs.timeplus.com/proton-kafka), so that we can process latest events in NATS continuously, in the streaming fashion, instead of running a bounded query for the Table Engine multiple times to get the updates.

**Comment by hwinkel** [comment]
> > Has not clickhouse already a NATS integration ?
> 
> no, clickhouse still does not fully support NATS Jetstream I guess? I find this PR still working in progress: [ClickHouse/ClickHouse#51281](https://github.com/ClickHouse/ClickHouse/pull/51281) full discussion here: [ClickHouse/ClickHouse#39459](https://github.com/ClickHouse/ClickHouse/issues/39459)

Hi,

OK; I see, Clickhouse has Core NATS but not acting as a Jetstream (durable) consumer.

**Comment by Sambit003** [comment]
/attempt #535

<div id="algora" />

<details>
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/cls13uc4c0002ih0g68ihcgc6/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by jovezhong** [comment]
Hi @Sambit003 , glad to see you pick up this task. Feel free to ask any questions here or in https://timeplus.com/slack

**Comment by Sambit003** [comment]
@jovezhong  why this is?? No support for GNU Compiler Collections?? 
![Screenshot from 2024-04-07 20-27-04](https://github.com/timeplus-io/proton/assets/58760654/25648ba3-e385-4e04-ae87-fd4797fb039e)

**Comment by yokofly** [comment]
> @jovezhong  why this is?? No support for GNU Compiler Collections?? 
> ![Screenshot from 2024-04-07 20-27-04](https://github.com/timeplus-io/proton/assets/58760654/25648ba3-e385-4e04-ae87-fd4797fb039e)
> 

yes . we do not support gcc, plz use clang16.
https://github.com/timeplus-io/proton/blob/develop/BUILD.md

historical reason, and we use forked llvm repo.
personally, llvm is more friendly for learning.

**Comment by ayewo** [comment]
I’d like to attempt this feature but probably wont be able to do anything substantive until next week or so.

/attempt #535

<div id="algora-attempt" />

| [Algora profile](https://console.algora.io/@/ayewo) | Completed bounties | Tech | Active attempts | Options |
| --- | --- | --- | --- | --- |
| @ayewo | &nbsp;&nbsp;&nbsp;3 timeplus bounties <br/> + 17 bounties from 4 projects | <div align="center">TypeScript, Rust, <br />Java & more</div> | <div align="center"></div> | [Cancel attempt](https://console.algora.io/api/bounties/cls13uc4c0002ih0g68ihcgc6/cancel-attempt) |

**Comment by Sambit003** [comment]
@ayewo hey, I'm working on this. So if I'm dropping it, I'll let you know. You can continue with my PR then.

**Comment by deependujha** [comment]
Hey, I'd love to work on this issue. Should  I continue?

**Comment by jovezhong** [comment]
Hi @deependujha , there were some GitHub users trying to work on this but I don't see PR for them since May. I will cancel those attempts and you can start working on them. It'll be C++ based to call NATS SDK to integrate with Timeplus Proton

**Comment by kruegernet** [comment]
@deependujha are you active on this?

**Comment by vishwamartur** [comment]
/attempt #535

<div id="algora-attempt" />

<details>
  <summary>Options</summary>
  <ul>
    <li>
      <a href="https://console.algora.io/api/bounties/cls13uc4c0002ih0g68ihcgc6/cancel-attempt">
        Cancel my attempt
      </a>
    </li>
  </ul>
</details>

**Comment by iho** [comment]
@jovezhong can I take it?

**Comment by jovezhong** [comment]
Hi @iho , please go ahead preparing the PR since I am not aware of other ongoing effort to add C++ based read/write capability in Timeplus yet. Feel free to ask questions here or join our community slack for more discussions https://timeplus.com/slack

**Comment by rishi-jat** [comment]
@mfreeman451 @jovezhong I would like to work on this issue is it still open to work ? Thanks!

**Comment by mfreeman451** [comment]
@rishi-jat as far as I know, no one has successfully completed this task, I assume the bounties are still available as well.

**Comment by chenziliang** [comment]
Hi @rishi-jat, thanks for your interest in contributing proton. The NATS support is still not done but the bounty program has been terminated.

**Comment by krapcys1-maker** [comment]
I’m looking into this and will propose a fix with tests shortly. Will update here if I hit any questions.

**Comment by Excellencedev** [comment]
Can I still work on this bounty ? @chenziliang @yokofly @gangtao

**Comment by chenziliang** [comment]
No.

**Comment by chenziliang** [comment]
User can use Timeplus Python table function to handle NATS data

**Comment by Shonke** [comment]
Hello, can I use Jetstream in Timeplus Proton (not Timeplus Enterprise) now?

**Comment by yokofly** [comment]
> Hello, can I use Jetstream in Timeplus Proton (not Timeplus Enterprise) now?

Yes, please try latest version https://github.com/timeplus-io/proton/releases/tag/v3.0.18

for usage: https://github.com/timeplus-io/proton/pull/1111#issue-3922533145
test case: https://github.com/timeplus-io/proton/tree/47b6ddf8654792eb5c5b84d06ca0cf70d14d7198/tests/cluster/smoke/0032_external_stream_nats
